grass/kernel.c

```diff
@@ -127,60 +127,51 @@ static void proc_yield() {
     proc_set_running(curr_pid);
 }
 
-struct pending_ipc *msg_buffer = (void*)(EGOS_HEAP_END);
-
-static int proc_try_send(struct syscall *sc, struct process *sender) {
-    if (msg_buffer->in_use == 1) return -1;
-    
+static int proc_try_send(struct process *sender) {
     for (uint i = 0; i < MAX_NPROCESS; i++) {
-        struct process dst = proc_set[i];
-        if (dst.pid == sc->msg.receiver && dst.status != PROC_UNUSED) {
+        struct process *dst = &proc_set[i];
+        if (dst->pid == sender->sc.msg.receiver && dst->status != PROC_UNUSED) {
             /* Destination is not receiving, or will not take msg from sender */
-            if (! (dst.status == PROC_PENDING_SYSCALL && dst.pending_syscall == SYS_RECV) )   return -1;
-            if (! (dst.receive_from == GPID_ALL || dst.receive_from == sender->pid) ) return -1;
+            if (! (dst->sc.type == SYS_RECV && !dst->sc.msg.received) ) return -1;
+            if (! (dst->sc.msg.sender == GPID_ALL || dst->sc.msg.sender == sender->pid) ) return -1;
             
-            msg_buffer->in_use = 1;
-            msg_buffer->sender = sender->pid;
-            msg_buffer->receiver = sc->msg.receiver;
-
-            memcpy(msg_buffer->msg, sc->msg.content, sizeof(sc->msg.content));
+            dst->sc.msg.sender   = sender->pid;
+            dst->sc.msg.received = 1;
+            memcpy(dst->sc.msg.content, sender->sc.msg.content, SYSCALL_MSG_LEN);
             return 0;
         }
     }
-    FATAL("proc_try_send: process %d sending to unknown process %d", sender->pid, sc->msg.receiver);
+    FATAL("proc_try_send: process %d sending to unknown process %d", sender->pid, sender->sc.msg.receiver);
 }
 
-static int proc_try_recv(struct syscall *sc, struct process *receiver) {
-    receiver->receive_from = sc->msg.sender;
+static int proc_try_recv(struct process *receiver) {
+    if (!receiver->sc.msg.received) return -1;
     
-    if (msg_buffer->in_use == 0 || msg_buffer->receiver != receiver->pid) return -1;
-
-    msg_buffer->in_use = 0;
-    sc->msg.sender = msg_buffer->sender;
-    memcpy(sc->msg.content, msg_buffer->msg, sizeof(sc->msg.content));
+    receiver->sc.msg.received = 0;
```

Comment:
I feel like you can remove this line, if you do the suggested modification in `excp_entry`.